common/archiver/s3store/visibilityArchiver.go

```diff
@@ -129,23 +132,30 @@ func (v *visibilityArchiver) Archive(
 		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeVisibilityRecord), tag.Error(err))
 		return err
 	}
-
-	// Upload archive to all indexes
-	key := constructTimestampIndex(URI.Path(), request.DomainID, request.WorkflowID, indexKeyCloseTimeout, request.CloseTimestamp, request.RunID)
-	if err := upload(ctx, v.s3cli, URI, key, encodedVisibilityRecord); err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
-		return err
+	var indexes = []struct {
+		primaryIndex            string
+		primaryIndexValue       string
+		secondaryIndex          string
+		secondaryIndexTimestamp int64
+	}{
+		{primaryIndexKeyWorkflowTypeName, request.WorkflowTypeName, secondaryIndexKeyCloseTimeout, request.CloseTimestamp},
+		{primaryIndexKeyWorkflowTypeName, request.WorkflowTypeName, secondaryIndexKeyStartTimeout, request.StartTimestamp},
+		{primaryIndexKeyWorkflowID, request.WorkflowID, secondaryIndexKeyCloseTimeout, request.CloseTimestamp},
+		{primaryIndexKeyWorkflowID, request.WorkflowID, secondaryIndexKeyStartTimeout, request.StartTimestamp},
 	}
-	key = constructTimestampIndex(URI.Path(), request.DomainID, request.WorkflowID, indexKeyStartTimeout, request.StartTimestamp, request.RunID)
-	if err := upload(ctx, v.s3cli, URI, key, encodedVisibilityRecord); err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
-		return err
+	// Upload archive to all indexes
+	for _, element := range indexes {
+		key := constructTimestampIndex(URI.Path(), request.DomainID, element.primaryIndex, element.primaryIndexValue, element.secondaryIndex, element.secondaryIndexTimestamp, request.RunID)
+		if err := upload(ctx, v.s3cli, URI, key, encodedVisibilityRecord); err != nil {
```

Comment:
Not related to this PR but do we want to add some retry logic here? Otherwise the caller will give up archiving and the visibility record will be lost.